feat: adding new `versions` package

arboleya opened the pull request:
# Summary

- Close #609
  - _Adds a new `versions` package_
  -  _Concentrates all important versions together_
  - _Handle `BUILD_VERSION` env variable on a single place_
  - _Provides all SDK-supported versions of the Fuel toolchain components_
  - _Provides CLI utility for comparing user-installed versions against the SDK ones_

> _More on the package [README](https://github.com/FuelLabs/fuels-ts/blob/aa/feat/adding-versions-package/packages/versions/README.md), and a few snippets bellow._

# Programmatic usage

It gives accurate versions of the Fuel toolchain supported by the SDK.

```ts
import { versions } from "@fuel-ts/versions";

const logger = new Logger(versions.FUELS);

console.log(versions);
// { FUELS: '0.21.2', FUEL_CORE: '0.14.0', FORC: '0.30.0' }
```

# CLI usage

Help users compare their local version against the supported ones.

```console
$ npx fuels versions
Supported fuel-core: 0.14.1
You're using fuel-core: 0.14.0

You can install/update them with:
 https://github.com/fuellabs/fuelup
```

QuinnLee commented:
In `provider.test.ts` there is a tests that does 

```
  it('can getVersion()', async () => {
    const provider = new Provider('http://127.0.0.1:4000/graphql');

    const version = await provider.getVersion();

    expect(version).toEqual('0.14.0');
  });
```

should we the hardcoded value `0.14.0` or use `version` instead?

arboleya commented:
> should we the hardcoded value `0.14.0` or use `version` instead?

I think this test can be safely deleted in favor of the one below. What do you think?

The version strings (_on the test below_) must be hardcoded inside the test.

https://github.com/FuelLabs/fuels-ts/blob/7449f47fddd8115b2f2770662afbbd010b6454ef/packages/versions/scripts/getToolchainVersions.test.ts#L17-L25

arboleya commented:
> should we the hardcoded value `0.14.0` or use `version` instead?

@QuinnLee  Good catch! Yes, I replaced the hardcoded version with the version's ref.

I also implemented all missing tests, including the one below, which covers version matching.

https://github.com/FuelLabs/fuels-ts/blob/7449f47fddd8115b2f2770662afbbd010b6454ef/packages/versions/scripts/getToolchainVersions.test.ts#L17-L25

luizstacio commented on README.md:
```diff
@@ -137,6 +138,21 @@ const contract = await factory.deployContract(factory);
 
+## Validating Toolchain Versions
+
+This utility can be used to check if you're using the supported versions.
```
Maybe add some more info here?

```suggestion
Fuels include a CLI utility that helps you check your installed toolchain version against the current SDK version installed.
```

arboleya commented on packages/versions/scripts/getToolchainVersions.ts:
```diff
@@ -0,0 +1,29 @@
+import { readFileSync } from 'fs';
+import { join } from 'path';
+
+export function getToolchainVersions() {
+  const root = join(__dirname, '..', '..', '..');
+  const dockerFilepath = join(root, 'services', 'fuel-core', 'Dockerfile');
+  const forcPackagePath = join(root, 'packages', 'forc-bin', 'package.json');
+  const fuelsPackagePath = join(root, 'packages', 'fuels', 'package.json');
+
+  const dockerContents = readFileSync(dockerFilepath, 'utf-8');
+  const forcJson = JSON.parse(readFileSync(forcPackagePath, 'utf-8'));
+  const fuelsJson = JSON.parse(readFileSync(fuelsPackagePath, 'utf-8'));
+
+  const dockerMatch = /fuel-core:v([0-9.]+)$/gm.exec(dockerContents);
+
+  if (!dockerMatch || dockerMatch.length < 2) {
+    throw new Error(`Could not read version from Dockerfile: \n ${dockerFilepath}`);
+  }
+
+  const FUEL_CORE = dockerMatch[1];
+  const FORC = forcJson.config.forcVersion;
+  const FUELS = fuelsJson.version;
+
+  return {
+    FUEL_CORE,
+    FORC,
+    FUELS,
+  };
+}
```
**Note to self**: Considering using the variables provided by:
https://github.com/FuelLabs/fuels-ts/blob/640af68ecadea46b159bd37310305623f2fabbf4/.github/workflows/release.yaml#L28-L30

arboleya commented on packages/versions/scripts/getToolchainVersions.ts:
```diff
@@ -0,0 +1,29 @@
+
+  const FUEL_CORE = dockerMatch[1];
+  const FORC = forcJson.config.forcVersion;
+  const FUELS = fuelsJson.version;
```
I was thinking about this one yesterday. Thanks for the insight!

I'll start using the versions from the [build env vars](https://github.com/FuelLabs/fuels-ts/pull/616#discussion_r1033739848) following your implementation on the [docs PR](https://github.com/FuelLabs/fuels-ts/pull/612). This should simplify everything and remove all the friction.

arboleya commented on packages/versions/src/cli.ts:
```diff
@@ -0,0 +1,47 @@
+  }
+
+  const forcOk = semver.satisfies(userForc, versions.FORC);
+  const coreOk = semver.satisfies(userFuelCore, versions.FUEL_CORE);
```
TBH, I'm unsure about this one. It might be too strict.

Should we use something else, like `server.lt()` or `server.minVersion()`?

arboleya replied:
I went with:
 - `semver.gt(userVersion, supportedVersion)` = `exit(0)` w/ caution message
 - `semver.eq(userVersion, supportedVersion)` = `exit(0)` w/ success message
 - otherwise fails = `exit(1)`](https://github.com/FuelLabs/fuels-
 
Here:
https://github.com/FuelLabs/fuels-ts/blob/70c795ee702dd323c6cac2b12d10c9353c972576/packages/versions/src/utils/compareUserVersions.ts#L13-L19

And here:
https://github.com/FuelLabs/fuels-ts/blob/9eb62672f40c3a1795d1624838cfd6a238c6cf8d/packages/versions/src/cli.ts#L40-L57

arboleya commented on packages/versions/src/index.ts:
```diff
@@ -0,0 +1,19 @@
+export const versions = {
+  FUELS: process.env.BUILD_VERSION || '0.21.2',
+  FUEL_CORE: '0.14.0',
+  FORC: '0.30.0',
```
Yes, especially after I do [this](https://github.com/FuelLabs/fuels-ts/pull/616#discussion_r1034490486), and remove the `prebuild` step.

It should end up looking like this:

```ts
export const versions = {
  FUELS: process.env.BUILD_VERSION || '~',
  FUEL_CORE: process.env.FUEL_CORE_VERSION || '~',
  FORC: process.env.FORC_VERSION || '~'
}
```

arboleya replied:
I chose to go with `0.0.0` because `~` is not a valid **_semver_**.

arboleya commented on packages/fuels/jest.config.ts:
```diff
@@ -0,0 +1,11 @@
+  testPathIgnorePatterns: ['/node_modules/', '/dist/'],
+  modulePathIgnorePatterns: ['/dist/'],
+  coveragePathIgnorePatterns: ['/dist/', '/test/', '.test.ts'],
+};
```
AFAIK, Jest has no concept of inheritance.

The ideal way is to have all configs separately and then use a macro-runner either via [`jest --projects`](https://jestjs.io/docs/configuration#projects-arraystring--projectconfig) or, preferably in our case, **_turborepo_** — _which should greatly speed up local development by leveraging [caching](https://turbo.build/repo/docs/core-concepts/caching) features_.

Although I can't go down this rabbit hole right now, I started configuring new packages with local configs in preparation for the future. The same is happening in parallel for the `abi-typegen`.

arboleya commented on packages/fuels/package.json:
```diff
@@ -44,11 +45,22 @@
-    "@fuel-ts/wordlists": "workspace:*"
+    "@fuel-ts/wallet": "workspace:*",
+    "@fuel-ts/wordlists": "workspace:*",
+    "commander": "^9.4.1"
```
It should not affect browsers in any way because `commander` is used only inside the `cli.ts` file, which is never imported by any other module except the app binary entry path (`bin.ts`); both of which are never referenced in any other place inside the package.

arboleya commented on packages/providers/package.json:
```diff
@@ -40,6 +40,7 @@
     "@fuel-ts/keystore": "workspace:*",
     "@fuel-ts/math": "workspace:*",
     "@fuel-ts/transactions": "workspace:*",
+    "@fuel-ts/versions": "workspace:*",
```
Of course, it is. NOT. 😅 

Eagles eye! 🦅

arboleya commented on packages/address/src/utils.ts:
```diff
@@ -10,10 +10,11 @@ import type {
 import { bech32m } from 'bech32';
 
-const logger = new Logger(process.env.BUILD_VERSION || '~');
+const logger = new Logger(versions.FUELS);
```
Makes sense. IIRC @pedronauck suggested the same on Slack.

We could also consider treating `logging` + `error handling` in conjunction.

But yes, I think in another issue / PR.

luizstacio commented on packages/fuels/src/cli.ts:
```diff
@@ -0,0 +1,23 @@
+    .description('checks for version incompatibilities')
+    .action(runVersions);
+
+  // program
```
Can you remove this comment here?
